File: php/src/Resource/Teams.php
<?php

declare(strict_types=1);

namespace Fivexer\SDK\Resource;

use Fivexer\SDK\Fivexer;
use Fivexer\SDK\Internal\Json;
use Fivexer\SDK\Model\PatchTeam;
use Fivexer\SDK\Model\Team;
use Fivexer\SDK\Model\TeamMembers;
use Fivexer\SDK\Model\TeamRoster;
use Fivexer\SDK\Model\UpsertTeam;

/**
 * Teams.
 *
 * A team is a routing primitive, not a label: the server derives a `tag` from the key and
 * matching sees that tag, so adding a worker to a team changes what work reaches them.
 */
final class Teams
{
    public function __construct(private readonly Fivexer $client)
    {
    }

    public function create(UpsertTeam $input): Team
    {
        return Team::fromArray($this->client->request('POST', '/teams', $input->toArray()) ?? []);
    }

    /** @return list<Team> */
    public function list(): array
    {
        $data = $this->client->request('GET', '/teams') ?? [];
        /** @var list<Team> */
        return Json::parseEach($data, 'teams', [Team::class, 'fromArray']);
    }

    public function get(string $teamId): Team
    {
        return Team::fromArray($this->client->request('GET', '/teams/' . \rawurlencode($teamId)) ?? []);
    }

    /** Partial update — the key is immutable. */
    public function patch(string $teamId, PatchTeam $patch): Team
    {
        return Team::fromArray(
            $this->client->request('PATCH', '/teams/' . \rawurlencode($teamId), $patch->toArray()) ?? []
        );
    }

    public function remove(string $teamId): void
    {
        $this->client->request('DELETE', '/teams/' . \rawurlencode($teamId));
    }

    public function members(string $teamId): TeamMembers
    {
        return TeamMembers::fromArray(
            $this->client->request('GET', '/teams/' . \rawurlencode($teamId) . '/members') ?? []
        );
    }

    /**
     * Replaces the roster wholesale — a worker absent from $workerIds is removed from the team.
     *
     * @param list<string> $workerIds
     */
    public function setMembers(string $teamId, array $workerIds): TeamRoster
    {
        return TeamRoster::fromArray(
            $this->client->request(
                'PUT',
                '/teams/' . \rawurlencode($teamId) . '/members',
                ['workerIds' => $workerIds]
            ) ?? []
        );
    }
}

File: java/src/main/java/io/fivexer/sdk/Teams.java
package io.fivexer.sdk;

import io.fivexer.sdk.internal.Json;
import io.fivexer.sdk.model.PatchTeam;
import io.fivexer.sdk.model.Team;
import io.fivexer.sdk.model.TeamList;
import io.fivexer.sdk.model.TeamMembers;
import io.fivexer.sdk.model.TeamRoster;
import io.fivexer.sdk.model.UpsertTeam;
import java.util.Collections;
import java.util.List;

/**
 * Teams.
 *
 * <p>A team is a routing primitive, not a label: the server derives a {@code tag} from the key
 * and matching sees that tag, so adding a worker to a team changes what work reaches them.
 */
public final class Teams {

    private final Fivexer client;

    Teams(Fivexer client) {
        this.client = client;
    }

    public Team create(UpsertTeam input) {
        return client.request("POST", "/teams", input.toJson(), null, Team.class, false);
    }

    public List<Team> list() {
        TeamList envelope = client.request("GET", "/teams", null, null, TeamList.class, false);
        return envelope.getTeams() == null ? Collections.emptyList() : envelope.getTeams();
    }

    public Team get(String teamId) {
        return client.request("GET", "/teams/" + Json.enc(teamId), null, null, Team.class, false);
    }

    /** Partial update — the {@code key} is immutable. */
    public Team patch(String teamId, PatchTeam patch) {
        return client.request("PATCH", "/teams/" + Json.enc(teamId), patch.toJson(), null,
                Team.class, false);
    }

    public void remove(String teamId) {
        client.request("DELETE", "/teams/" + Json.enc(teamId), null, null, Void.class, true);
    }

    public TeamMembers members(String teamId) {
        return client.request("GET", "/teams/" + Json.enc(teamId) + "/members", null, null,
                TeamMembers.class, false);
    }

    /** Replaces the roster wholesale — a worker absent from {@code workerIds} is removed. */
    public TeamRoster setMembers(String teamId, List<String> workerIds) {
        return client.request("PUT", "/teams/" + Json.enc(teamId) + "/members",
                Json.object("workerIds", workerIds), null, TeamRoster.class, false);
    }
}
